packages/@aws-cdk/aws-appmesh/lib/virtual-listener.ts

```diff
@@ -0,0 +1,338 @@
+   * @default - HTTP
+   */
+  readonly protocol?: Protocol;
+  /**
+   * Timeout for HTTP protocol
+   *
+   * @default - None
+   */
+  readonly timeout?: HttpTimeout;
+}
+
+/**
+ * Represent the GRPC Node Listener prorperty
+ */
+export interface GrpcNodeListenerProps extends VirtualNodeListenerProps {
+  /**
+   * Timeout for GRPC protocol
+   *
+   * @default - None
+   */
+  readonly timeout?: GrpcTimeout;
+}
+
+/**
+ * Represent the TCP Node Listener prorperty
+ */
+export interface TcpNodeListenerProps extends VirtualNodeListenerProps {
+  /**
+   * Timeout for TCP protocol
+   *
+   * @default - None
+   */
+  readonly timeout?: TcpTimeout;
+}
+
+/**
+ * Returns an HealthCheck for a VirtualNode
+ */
+function renderHealthCheck(pm: PortMapping, hc: HealthCheck | undefined): CfnVirtualNode.HealthCheckProperty | undefined {
+  /**
+   * Minimum and maximum thresholds for HeathCheck numeric properties
+   *
+   * @see https://docs.aws.amazon.com/app-mesh/latest/APIReference/API_HealthCheckPolicy.html
+   */
+  const HEALTH_CHECK_PROPERTY_THRESHOLDS: {[key in (keyof CfnVirtualNode.HealthCheckProperty)]?: [number, number]} = {
+    healthyThreshold: [2, 10],
+    intervalMillis: [5000, 300000],
+    port: [1, 65535],
+    timeoutMillis: [2000, 60000],
+    unhealthyThreshold: [2, 10],
+  };
+
+  if (hc === undefined) { return undefined; }
+
+  if (hc.protocol === Protocol.TCP && hc.path) {
+    throw new Error('The path property cannot be set with Protocol.TCP');
+  }
+
+  if (hc.protocol === Protocol.GRPC && hc.path) {
+    throw new Error('The path property cannot be set with Protocol.GRPC');
+  }
+
+  const healthCheck: CfnVirtualNode.HealthCheckProperty = {
+    healthyThreshold: hc.healthyThreshold || 2,
+    intervalMillis: (hc.interval || cdk.Duration.seconds(5)).toMilliseconds(), // min
+    path: hc.path || (hc.protocol === Protocol.HTTP ? '/' : undefined),
+    port: hc.port || pm.port,
+    protocol: hc.protocol || pm.protocol,
+    timeoutMillis: (hc.timeout || cdk.Duration.seconds(2)).toMilliseconds(),
+    unhealthyThreshold: hc.unhealthyThreshold || 2,
+  };
+
+  (Object.keys(healthCheck) as Array<keyof CfnVirtualNode.HealthCheckProperty>)
+    .filter((key) =>
+      HEALTH_CHECK_PROPERTY_THRESHOLDS[key] &&
+          typeof healthCheck[key] === 'number' &&
+          !cdk.Token.isUnresolved(healthCheck[key]),
+    ).map((key) => {
+      const [min, max] = HEALTH_CHECK_PROPERTY_THRESHOLDS[key]!;
+      const value = healthCheck[key]!;
+
+      if (value < min) {
+        throw new Error(`The value of '${key}' is below the minimum threshold (expected >=${min}, got ${value})`);
+      }
+      if (value > max) {
+        throw new Error(`The value of '${key}' is above the maximum threshold (expected <=${max}, got ${value})`);
+      }
+    });
+
+  return healthCheck;
+}
+
+/**
+ * Returns the ListenerTimeoutProperty for HTTP protocol
+ */
+function renderTimeout(tm: any, pr: Protocol): CfnVirtualNode.ListenerTimeoutProperty {
```

Comment:
Can this be a protected function of `VirtualNodeListener`? That way you can just reference the class properties and not pass parameters to this function